### PR TITLE
feat: add manual inject styles function into the project

### PR DESCRIPTION
**Pull request by danielbarion**
idea from #1025

### CODE REVIEW

**Comment by danielbarion** [comment]
I created the flow to inject the styles manually from the package instead of letting Rollup handle it by itself.

what is pending?
* add a switch to enable and disable styles injection.

Suggestion: 
In the current way that this was implemented, we can add the switch by environment variables, ex:
`REACT_TOOLTIP_DISABLE_STYLES=true`

But this doesn't sound like the best way to handle this switch.

So, I need help here :)

File: src/utils/handle-style.ts
Comment (gabrieljablonski) [inline src/utils/handle-style.ts:44]: Please test if something like this works (too lazy to setup and test it myself 😑)

The idea is to use `import { removeStyle } from 'react-tooltip'`

File: src/index.tsx
Comment (gabrieljablonski) [inline src/index.tsx:22]: I didn't mean to commit it, but why was `@ts-ignore` being used here? No warnings or errors on my environment.

File: src/utils/handle-style.ts
Comment (gabrieljablonski) [inline src/utils/handle-style.ts:43]: Also if you make any changes, do this one as well, forgot to clean it up.

```suggestion
  const style = document.getElementById(REACT_TOOLTIP_STYLES_ID)
  style?.remove()
```

**Comment by gabrieljablonski** [comment]
How does removing the style injection affect the tooltip functionally? Since it depends on the `.tooltip` and `.show` classes to work properly, will the tooltip still show and hide without any issues?

**Comment by danielbarion** [comment]
> How does removing the style injection affect the tooltip functionally? Since it depends on the `.tooltip` and `.show` classes to work properly, will the tooltip still show and hide without any issues?

This will probably break the tooltip feature, one thing that we can do is separate it into 2 CSS files:
1. for basic functionality - keep the tooltip working as expected
2. for styling - without this one, we will only show and hide the tooltip content without any styling

please let me know your thoughts about it

File: src/index.tsx
Comment (danielbarion) [reply src/index.tsx:22]: yeap, all good! thanks!

**Comment by gabrieljablonski** [comment]
@danielbarion Having a separate injection for functional styling seems like an adequate solution.

My suggestion (make sure to check if I missed something):

```css
/* react-tooltip-core.css (feel free to choose a better name for the file) */
.tooltip {
  position: absolute;
  visibility: hidden;
  opacity: 0;
}

.fixed {
  position: fixed;
}

.arrow {
  position: absolute;
}

.noArrow {
  display: none;
}

.show {
  visibility: visible;
  opacity: var(--rt-opacity);
}
```

`.fixed` and `.noArrow` could be left out. I included them just so the `positionStrategy` and `noArrow` props still work. Your call if we should consider them functional styling or not.

**Comment by danielbarion** [comment]
Beta version released with the last commit:

```
yarn add react-tooltip@5.15.0-beta.1041.0
```
or
```
npm install react-tooltip@5.15.0-beta.1041.0
```

I only tested without the styles, I'll test everything later, but feel free to test it or update anything.

Thanks!

File: src/components/Tooltip/Tooltip.tsx
Comment (gabrieljablonski) [inline src/components/Tooltip/Tooltip.tsx:583]: Is keeping both styles intentional?

**Comment by danielbarion** [comment]
Yes, if you check, they has different attributes on their files (if I'm not
wrong)

I will run more tests, I had to leave to computer, so, this is just to
validate the idea

Em sex., 16 de jun. de 2023 15:53, Gabriel Jablonski <
***@***.***> escreveu:

File: src/components/Tooltip/Tooltip.tsx
Comment (danielbarion) [reply src/components/Tooltip/Tooltip.tsx:583]: yes, but please feel free to reorganize the styles, I tried to keep the necessary CSS for position and basic behavior into the Core file, I'll check again this week when I have more time, but please, go ahead if you want to change anything

**Comment by danielbarion** [comment]
https://github.com/ReactTooltip/react-tooltip/assets/9615850/22de8b3d-59f3-4b9b-9342-3ac9e4ab4051

@gabrieljablonski 🔝

**Comment by gabrieljablonski** [comment]
> The code can be used in a button or inside of a `useEffect`

Does calling it globally (outside any components) work? That would be the most convenient IMO

**Comment by danielbarion** [comment]
@gabrieljablonski I'm not totally sure that I got how to test it on Next 13, but it breaks because I'm importing from a custom export:

```
"use client"

export * from 'react-tooltip'
```

and importing this function from the package also breaks it too.

Can you test it or guide me on how to test it, please?

My test environment is just a fresh Next 13 project and I'm using Yalc to test the implementation, but you can try the beta release from NPM instead of Yalc

**Comment by danielbarion** [comment]
@gabrieljablonski another option is:
* let the guys import the `styles` (basically the colors - content from styles.module.css) and we only inject the core styles by default

**Comment by danielbarion** [comment]
> If we do this we're back to how it was before, which would not be ideal IMO.

I do agree, but now we have the basic feature of the Tooltip with the Core styles before nothing was working because of 0 styles of tooltip on the page. I'm just thinking of projects that don't want to inject the colors from react-tooltip and this can cause re-paint of the elements in the page (probably this impact will be minimal, but I'm thinking if it's possible to reduce it to 0).

Does it make sense?

**Comment by gabrieljablonski** [comment]
Since most users will probably just use the default styling and partially override the styling, I really like the idea of not having to import something to apply the styles.

I'm fine with the env variable approach, but then we should consider how useful still having `removeStyle()` would be. Would anyone really want to remove the default styling on the fly? So having just the one way to do it feels a lot cleaner.

**Comment by danielbarion** [comment]
Keeping the `removeStyles` is great for simple projects using pure HTML (importing React from CDN), or projects using other languages on the server side (like Laravel projects or something like that).

I don't know, I'm probably overthinking here, if you believe it's not useful to keep the remove styles function, I'll agree with you and all good :)

**Comment by danielbarion** [comment]
@gabrieljablonski the `force push` was because I did `git pull -r origin master`, after the rebase, I had to push -f

**Review by gabrieljablonski — APPROVED** [review]

**Comment by danielbarion** [comment]
```
yarn add react-tooltip@5.16.0-beta.1041.0
```
or
```
npm install react-tooltip@5.16.0-beta.1041.0
```

Check the PR changes for more information.

**Comment by dsternlicht** [comment]
@danielbarion any idea when this is going to be merged officially? :)

**Comment by gabrieljablonski** [comment]
@dsternlicht just working out the final kinks. An official version should be released in a couple days.

Until then, don't worry about using the beta version, it should work just fine.

**Comment by dsternlicht** [comment]
> @dsternlicht just working out the final kinks. An official version should be released in a couple days.
> 
> Until then, don't worry about using the beta version, it should work just fine.

Sounds good. Thanks @gabrieljablonski!

**Comment by danielbarion** [comment]
@gabrieljablonski I tested again and I did two new small commits, please check them before we merge this PR

thanks!

**Review by gabrieljablonski — APPROVED** [review]

**Comment by dsternlicht** [comment]
🎉

**Comment by danielbarion** [comment]
available at [v5.16.0](https://github.com/ReactTooltip/react-tooltip/releases/tag/v5.16.0)

**Pull request merged**
